src/Lykke.Service.CandlesHistory.Services/Candles/RedisCandlesCacheService.cs

```diff
@@ -0,0 +1,171 @@
+using Lykke.Job.CandlesProducer.Contract;
+using Lykke.Service.CandlesHistory.Core.Domain.Candles;
+using Lykke.Service.CandlesHistory.Core.Services.Candles;
+using MoreLinq;
+using StackExchange.Redis;
+
+namespace Lykke.Service.CandlesHistory.Services.Candles
+{
+    /// <summary>
+    /// Caches candles in the redis using lexographical indexes with candles data as the auxiliary information of the index
+    /// </summary>
+    [UsedImplicitly]
+    public class RedisCandlesCacheService : ICandlesCacheService
+    {
+        private const string TimestampFormat = "yyyyMMddHHmmss";
+
+        private readonly IDatabase _database;
+        private readonly MarketType _market;
+        private readonly int _amountOfCandlesToStore;
+
+        public RedisCandlesCacheService(IDatabase database, MarketType market, int amountOfCandlesToStore)
+        {
+            _database = database;
+            _market = market;
+            _amountOfCandlesToStore = amountOfCandlesToStore;
+        }
+
+        public IImmutableDictionary<string, IImmutableList<ICandle>> GetState()
+        {
+            throw new NotSupportedException();
+        }
+
+        public void SetState(IImmutableDictionary<string, IImmutableList<ICandle>> state)
+        {
+            throw new NotSupportedException();
+        }
+
+        public string DescribeState(IImmutableDictionary<string, IImmutableList<ICandle>> state)
+        {
+            throw new NotSupportedException();
+        }
+
+        public async Task InitializeAsync(
+            string assetPairId, 
+            CandlePriceType priceType,
+            CandleTimeInterval timeInterval,
+            IReadOnlyCollection<ICandle> candles)
+        {
+            foreach (var candle in candles)
+            {
+                if (candle.AssetPairId != assetPairId)
+                {
+                    throw new ArgumentException($"Candle {candle.ToJson()} has invalid AssetPriceId", nameof(candles));
+                }
+                if (candle.PriceType != priceType)
+                {
+                    throw new ArgumentException($"Candle {candle.ToJson()} has invalid PriceType", nameof(candles));
+                }
+                if (candle.TimeInterval != timeInterval)
+                {
+                    throw new ArgumentException($"Candle {candle.ToJson()} has invalid TimeInterval", nameof(candles));
+                }
+            }
+
+            // TODO: This is non concurrent-safe update
+
+            var key = GetKey(assetPairId, priceType, timeInterval);
+
+            // Cleans cache
+
+            await _database.SortedSetRemoveRangeByRankAsync(key, 0, -1);
+
+            foreach (var candlesBatch in candles.Batch(100))
+            {
+                var entites = candlesBatch
+                    .Select(candle => new SortedSetEntry(SerializeCandle(candle), 0))
+                    .ToArray();
+
+                await _database.SortedSetAddAsync(key, entites);
```

Comment:
Use redit.Batch. The example https://github.com/LykkeCity/Lykke.Service.FixGateway/blob/dev/src/Lykke.Service.FixGateway.Services/ClientOrderIdProvider.cs#L45

Reply:
I would add some reasonable TTL

Reply:
Such an initialization is still used only for purposes of an old unit test. Please, see the previos comment.